source/slang/emit.cpp

```diff
@@ -3599,34 +3601,58 @@ struct EmitVisitor
         }
     }
 
-    void emitComparison(EmitContext* ctx, IRInst* inst, IREmitMode mode, EOpInfo& inOutOuterPrec, const EOpInfo& opPrec, bool* needCloseOut)
+    void _maybeEmitGLSLVectorCast(EmitContext* ctx, IRType* vecType, IRInst* inst, IREmitMode mode, const EOpInfo& opPrec)
     {
-        *needCloseOut = maybeEmitParens(inOutOuterPrec, opPrec);
-
-        if (getTarget(ctx) == CodeGenTarget::GLSL
-            && as<IRVectorType>(inst->getOperand(0)->getDataType())
-            && as<IRVectorType>(inst->getOperand(1)->getDataType()))
+        if (as<IRVectorType>(inst->getDataType()) == nullptr)
         {
-            const char* funcName = getGLSLVectorCompareFunctionName(inst->op);
-            SLANG_ASSERT(funcName);
-
-            emit(funcName);
+            emitIRType(ctx, vecType);
             emit("(");
-            emitIROperand(ctx, inst->getOperand(0), mode, leftSide(inOutOuterPrec, opPrec));
-            emit(",");
-            emitIROperand(ctx, inst->getOperand(1), mode, rightSide(inOutOuterPrec, opPrec));
+            emitIROperand(ctx, inst, mode, kEOp_General);
             emit(")");
         }
         else
         {
-            emitIROperand(ctx, inst->getOperand(0), mode, leftSide(inOutOuterPrec, opPrec));
-            emit(" ");
-            emit(opPrec.op);
-            emit(" ");
-            emitIROperand(ctx, inst->getOperand(1), mode, rightSide(inOutOuterPrec, opPrec));
+            emitIROperand(ctx, inst, mode, opPrec);
         }
     }
 
+    void emitComparison(EmitContext* ctx, IRInst* inst, IREmitMode mode, EOpInfo& inOutOuterPrec, const EOpInfo& opPrec, bool* needCloseOut)
+    {
+        *needCloseOut = maybeEmitParens(inOutOuterPrec, opPrec);
+
+        if (getTarget(ctx) == CodeGenTarget::GLSL)
+        {
+            IRInst* left = inst->getOperand(0);
+            IRInst* right = inst->getOperand(1);
+
+            auto leftVectorType = as<IRVectorType>(left->getDataType());
+            auto rightVectorType = as<IRVectorType>(right->getDataType());
+
+            // If either side is a vector handle as a vector
+            if (leftVectorType || rightVectorType)
+            {
+                const char* funcName = getGLSLVectorCompareFunctionName(inst->op);
+                SLANG_ASSERT(funcName);
+
+                const auto vecType = leftVectorType ? leftVectorType : rightVectorType;
+                
+                emit(funcName);
+                emit("(");
+                _maybeEmitGLSLVectorCast(ctx, vecType, left, mode, leftSide(inOutOuterPrec, opPrec));
```

Comment:
Passing in `leftSide(inOutOuterPrec, opPrec)` here (and the equivalent for the other argument) shouldn't be needed. These should be able to use `kEOp_General`.

The basic idea of the parentheses-insertion system is that for every expression we are going to emit we pass down the precedence of its surrounding context, which you can think of as encoding the operators to the immediate left and right. So if we had a context like:

```
a + [...] * b
```

and we were about to emit code to fill in the `[...]` part, then the precedence info would tell us that there is a `+` to our left, and a `*` to our right. If the expression we are going to emit into the `[...]` turns out to be `[...] * [...]`, then we compute the precedence information for `*`, and it tells us we don't need to parenthesize the sub-expression.

Then before we recurse into the left and right sides of `[...] * [...]` we need to compute the surrounding precedence for the left and right sides to use. That is what the `leftSide` and `rightSide` functions do: they take the current outer precedence (e.g., "`+` on the left, and `*` on the right") and the precedence info for the op we are emitting now (a `*`) and compute the precedence to use for the next sub-expression down (e.g., on the left side we have `+` on the left and `*` on the right again, while on the right we have `*` on the left and the right).

In cases where the context for a sub-expression doesn't depend on the surrounding context, we can pass in the precedence it should use directly (e.g., the use of `kEOp_General` for places where we are dealing with comma-separated argument lists).